src/WriteStream.h

```diff
@@ -78,35 +78,63 @@ class WriteStream {
         writeNextTuple(tuple.data);
     }
 
-    void writeValue(std::ostream& os, size_t col, RamDomain repr) {
-        switch (kindMask.at(col)) {
-            case 'i':
+    void writeValue(std::ostream& os, RamDomain repr, int typeId) {
+        Kind kind = typeTable.getKind(typeId);
+        switch (kind) {
+            case Kind::NUMBER:
                 os << repr;
                 break;
 
-            case 's':
+            case Kind::SYMBOL:
                 os << symbolTable.unsafeResolve(repr);
                 break;
 
-            case 'r': {
-                int arity = recordArityMask.at(col);
-                const auto& record = recordTable.getRecord(arity, repr);
-
-                os << "UnnamedRecord[" << join(record, ",") << "]";
+            case Kind::RECORD: {
+                // special case: nil
+                // TODO: assuming nil = 0
+                if (repr == 0) {
+                    os << "nil";
+                    break;
+                }
+
+                // get record metadata
+                const auto& name = typeTable.getName(typeId);
+                const auto& fieldTypes = typeTable.getFieldTypes(typeId);
+                int arity = fieldTypes.size();
+
+                // get record data
+                const auto& record = recordTable.getRecord(arity + 1, repr);
+
+                // print out the record recursively
+                os << name << "[";
+                for (int i = 0; i < arity; i++) {
```

Comment:
```suggestion
                for (size_t i = 0; i < arity; i++) {
```
size_t since we're comparing sizes